### ISSUE TITLE
Allow for constraints to take into account job allocations from another job

### ISSUE THREAD

**Issue by benvanstaveren** [hashicorp/nomad#11066]
### Proposal
It would be nice to have the ability to have "dynamic" constraints on jobs - I guess it would be dynamic, can't really come up with a better term. Mostly just that a constraint should exist that can take into account the existance (or absence) of a running/healthy allocation for another job.

### Use-cases

We run a bit of software that we offer as SaaS and for scaling's sake, it requires some databases to run - at the moment we solve this by having the database (hsqldb) run as a sidecar to the actual job, but this gives us some overhead, and for data synchronisation it requires a bit of a jump-through-hoops approach that, while it works, is ugly as sin. 

The easiest solution in our eyes is to run the hsqldb as a system job so each physical node runs one, so the whole pipeline becomes much easier. However, we can't run our SaaS containers if, for some reason, a node does not have the hsqldb allocation running (and system jobs do not seem to be evaluated automatically on a regular basis). It would be great to be able to say in a constraint that a job can only be placed on a node where another job has an active (and healthy) allocation.

**Comment by lgfa29** [comment]
Thanks for the idea @benvanstaveren, we will discuss more about this 🙂

**Comment by benvanstaveren** [comment]
Cheers! I know this probably won't be the easiest thing to figure out, but if it would be possible that'd be awesome.

**Comment by alexiri** [comment]
This sounds like it could be done with #11197 as well. @benvanstaveren, would that allow you to do what you want?

**Comment by benvanstaveren** [comment]
@alexiri not entirely, because my case requires the ability to have a constraint on whether or not an allocation is present on a physical node which puts it more into a "what nodes are eligible for this job" area...

**Comment by alexiri** [comment]
I thought it could be something like this:
```
job "hsqldb" {
  meta {
    job_class = "hsqldb"
  }
  ...
}

job "frontend" {
  # Only run this job where a hsqldb job is already running
  constraint {
    attribute = "${node.meta.job_class.hsqldb}"
    operator  = "="
    value     = "1"
  }
  ...
}
```

**Comment by benvanstaveren** [comment]
@alexiri yes, then it would work, I do think it would be confusing to have it as node.meta.job_class etc. since it's not a node-defined meta attribute, maybe better as "${job.meta.job_class.hsqldb}" or something, but I guess that's more for #11197 :D

**Comment by tgross** [comment]
See also: https://github.com/hashicorp/nomad/issues/24610
